source/agora/common/Hash.d

```diff
@@ -0,0 +1,159 @@
+/*******************************************************************************
+
+    Function definition and helper related to hashing
+
+    The actual definition of the `Hash` type is in `agora.common.Data`,
+    as parts of the system might pass along `Hash` without having to know
+    what / how they were produced.
+
+    However, this module expose functionalities for modules that want to do
+    hashing. The interface is designed so that this module knows about the
+    hash types, and aggregates implementing the interface only deal with their
+    members.
+
+    Due to a language limitation (one can't overload based on return value),
+    this module expose two main functions:
+    - `hashFull(T)`, which returns a `Hash`
+    - `hashPart(T, HasherDg)` which does not return anything and should be used
+      to accumulate data to hash.
+
+    For safety reason, a data structure currently need to explicitly define
+    a `hash` function to support hashing.
+    This limitation might be lifted in the future, but a few things
+    need to be taken into account:
+    - Alignment / packing: We can't have unaligned / unpacked structures
+      as it would create malleability issues
+    - Non-public members: We can't deal with anything non-public (we can use
+      `.tupleof` but it's impossible to tell the usage of the member,
+      e.g. it could be a cache of the hash)
+    - Indirections / references types
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Hash;
+
+import agora.common.Data;
+
+import libsodium;
+
+
+///
+nothrow @nogc @safe unittest
```

Comment:
It's a little weird to put the unittest first and then the implementation?

Reply:
Module documentation